src/qutip_qip/circuit/mat_renderer.py

```diff
@@ -0,0 +1,1025 @@
+            The list of wires the gate is acting on (control and target).
+
+        layer : int
+            The layer the gate is acting on.
+        """
+
+        xskip = []
+        if self.style.align_layer:
+            wire_list = list(range(self._qwires))
+
+        for wire in wire_list:
+            xskip.append(sum(self._layer_list[wire][:layer]))
+
+        return max(xskip)
+
+    def _get_text_width(
+        self,
+        text: str,
+        fontsize: float,
+        fontweight: Union[float, str],
+        fontfamily: str,
+        fontstyle: str,
+    ) -> float:
+        """
+        Get the width of the text to be plotted.
+
+        Parameters
+        ----------
+        text : str
+            The text to be plotted.
+
+        fontsize : float
+            The fontsize of the text.
+
+        fontweight : str or float
+            The fontweight of the text.
+
+        fontfamily : str
+            The fontfamily of the text.
+
+        fontstyle : str
+            The fontstyle of the text.
+        Returns
+        -------
+        float
+            The width of the text in cm.
+        """
+
+        text_obj = plt.Text(
+            0,
+            0,
+            text,
+            fontsize=fontsize,
+            fontweight=fontweight,
+            fontfamily=fontfamily,
+            fontstyle=fontstyle,
+        )
+        self._ax.add_artist(text_obj)
+
+        bbox = text_obj.get_window_extent(
+            renderer=self._ax.figure.canvas.get_renderer()
+        )
+        inv = self._ax.transData.inverted()
+        bbox_data = bbox.transformed(inv)
+        text_obj.remove()
+
+        return bbox_data.width * 2.54 * 3
+
+    def _manage_layers(
+        self,
+        gate_width: float,
+        wire_list: List[int],
+        layer: int,
+        xskip: float = 0,
+    ) -> None:
+        """
+        Manages and updates the layer widths according to the gate's width just plotted.
+
+        Parameters
+        ----------
+        gate_width : float
+            The width of the gate to be plotted.
+
+        wire_list : list
+            The list of wires the gate is acting on (control and target).
+
+        layer : int
+            The layer the gate is acting on.
+
+        xskip : float, optional
+            The horizontal value for getting to requested layer. The default is 0.
+        """
+
+        for wire in wire_list:
+            if len(self._layer_list[wire]) > layer:
+                if (
+                    self._layer_list[wire][layer]
+                    < gate_width + self.style.gate_margin * 2
+                ):
+                    self._layer_list[wire][layer] = (
+                        gate_width + self.style.gate_margin * 2
+                    )
+            else:
+                temp = xskip - sum(self._layer_list[wire]) if xskip != 0 else 0
+                self._layer_list[wire].append(
+                    temp + gate_width + self.style.gate_margin * 2
+                )
```

Comment:
Maybe leave a few words just briefly commenting on what those if-else branch corresponds to, i.e., the current layer is full / if not full what will happen.